### PR TITLE
WIP: Use contract bindings for Genesis contracts

### PR DESCRIPTION
**Pull request by Raneet10**
This PR introduces the use of go bindings for genesis contracts in the codebase, which can be effectively used in e2e tests and can come handy during debugging. This is still a WIP

Ref: [Task](https://polygon.atlassian.net/jira/software/projects/POS/boards/1?selectedIssue=POS-603)

### CODE REVIEW

File: consensus/bor/contract/client.go [hunk 1]
```diff
@@ -1,66 +1,51 @@
...
-	stateReceiverABI = `[{"constant":true,"inputs":[],"name":"SYSTEM_ADDRESS","outputs":[{"internalType":"address","name":"","type":"address"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":true,"inputs":[],"name":"lastStateId","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":false,"inputs":[{"internalType":"uint256","name":"syncTime","type":"uint256"},{"internalType":"bytes","name":"recordBytes","type":"bytes"}],"name":"commitState","outputs":[{"internalType":"bool","name":"success","type":"bool"}],"payable":false,"stateMutability":"nonpayable","type":"function"}]`
-)
+func NewGenesisContractsClient(chainConfig *params.ChainConfig, validatorSetAddr, stateReceiverAddr common.Address) *GenesisContractsClient {
+	backend, err := ethclient.Dial("127.0.0.1")
+	if err != nil {
+		log.Error("Failed to initialize ethereum client", "error", err)
```
Comment (JekaMas) [inline consensus/bor/contract/client.go:31]: could we return this errors and check them then?

File: consensus/bor/contract/client.go [hunk 2]
```diff
@@ -77,58 +62,37 @@ func (gc *GenesisContractsClient) CommitState(
...
-
 	t := event.Time.Unix()
 
-	data, err := gc.stateReceiverABI.Pack(method, big.NewInt(0).SetInt64(t), recordBytes)
-	if err != nil {
-		log.Error("Unable to pack tx for commitState", "error", err)
-		return 0, err
+	opts := &bind.TransactOpts{
+		From:     types.SystemAddress,
+		GasPrice: big.NewInt(0),
```
Comment (JekaMas) [inline consensus/bor/contract/client.go:69]: `SetInt64(t)` was lost?

File: consensus/bor/heimdall/span/spanner.go [hunk 3]
```diff
@@ -3,82 +3,57 @@ package span
...
-	ethAPI                   api.Caller
-	validatorSet             abi.ABI
+	validatorSet             *contracts.BorValidatorSet
 	chainConfig              *params.ChainConfig
 	validatorContractAddress common.Address
 }
 
-func NewChainSpanner(ethAPI api.Caller, validatorSet abi.ABI, chainConfig *params.ChainConfig, validatorContractAddress common.Address) *ChainSpanner {
+func NewChainSpanner(chainConfig *params.ChainConfig, validatorContractAddress common.Address) (*ChainSpanner, error) {
+	backend, err := ethclient.Dial("127.0.0.1")
```
Comment (JekaMas) [inline consensus/bor/heimdall/span/spanner.go:29]: why do we need it?

File: consensus/bor/heimdall/span/spanner.go [hunk 4]
```diff
@@ -3,82 +3,57 @@ package span
...
 
 type ChainSpanner struct {
-	ethAPI                   api.Caller
-	validatorSet             abi.ABI
+	validatorSet             *contracts.BorValidatorSet
 	chainConfig              *params.ChainConfig
 	validatorContractAddress common.Address
 }
 
-func NewChainSpanner(ethAPI api.Caller, validatorSet abi.ABI, chainConfig *params.ChainConfig, validatorContractAddress common.Address) *ChainSpanner {
```
Comment (JekaMas) [inline consensus/bor/heimdall/span/spanner.go:32]: could we keep interfaces like `ethAPI api.Caller` or introduce new ones? I wouldn't expect to make an http connection inside such constructors.

Comment (JekaMas) [reply consensus/bor/heimdall/span/spanner.go:32]: like `backend bind.ContractBackend`

**Comment by marcello33** [comment]
I'd keep the `.sol` file to make the bindings dynamic
Also, can we write a simple test to make sure everything works as expected when it comes to contract instance and its functioning?

**Comment by marcello33** [comment]
Please make sure the code compiles

File: accounts/abi/bind/base_test.go [hunk 5]
```diff
@@ -70,7 +73,7 @@ func (mt *mockTransactor) EstimateGas(ctx context.Context, call ethereum.CallMsg
 	return 0, nil
 }
 
-func (mt *mockTransactor) SendTransaction(ctx context.Context, tx *types.Transaction) error {
+func (mt *mockTransactor) SendTransaction(ctx context.Context, state *state.StateDB, header *types.Header, chainContext core.ChainContext, isSystem bool, chainConfig *params.ChainConfig, tx *types.Transaction) error {
```
Comment (JekaMas) [inline accounts/abi/bind/base_test.go:76]: we shouldn't change this interfaces. could we do the thing without it?

**Pull request closed without merging**